Sources/AblyChat/ChatAPI.swift

```diff
@@ -0,0 +1,232 @@
+            continuation.resume(throwing: ARTErrorInfo.createUnknownError())
+            return
+        }
+        
+        do {
+            let jsonData = try JSONSerialization.data(withJSONObject: paginatedResponse.items)
+            let decodedData = try JSONDecoder().decode([RES].self, from: jsonData)
+            let result = paginatedResponse.toPaginatedResult(items: decodedData)
+            continuation.resume(returning: result)
+        } catch {
+            continuation.resume(throwing: error)
+        }
+    }
+}
+
+extension ARTHTTPPaginatedResponse {
+    func toPaginatedResult<T: Codable & Sendable>(items: [T]) -> PaginatedResultWrapper<T>{
+        
+        return PaginatedResultWrapper(paginatedResponse: self, items: items)
+    }
+}
+
+public final class PaginatedResultWrapper<T: Codable & Sendable>: PaginatedResult {
+    public typealias T = T
+    
+    public let items: [T]
+    public let hasNext: Bool
+    public let isLast: Bool
+    private let paginatedResponse: ARTHTTPPaginatedResponse
+    
+    public init(paginatedResponse: ARTHTTPPaginatedResponse, items: [T]) {
+        self.items = items
+        self.hasNext = paginatedResponse.hasNext
+        self.isLast = paginatedResponse.isLast
+        self.paginatedResponse = paginatedResponse
+    }
+    
+    /// Asynchronously fetch the next page if available
+    public var next: (any PaginatedResult<T>)? {
+        get async throws {
+            if #available(macOS 13.0.0, *) {
+                return try await withCheckedThrowingContinuation { continuation in
+                    paginatedResponse.next { paginatedResponse, error in
+                        ARTHTTPPaginatedCallbackWrapper(callback: (paginatedResponse, error)).handleResponse(continuation: continuation)
+                    }
+                }
+            } else {
+                return nil
+            }
+        }
+    }
+    
+    /// Asynchronously fetch the first page
+    public var first: any PaginatedResult<T> {
+        get async throws {
+            if #available(macOS 13.0.0, *) {
+                return try await withCheckedThrowingContinuation { continuation in
+                    paginatedResponse.first { paginatedResponse, error in
+                        ARTHTTPPaginatedCallbackWrapper(callback: (paginatedResponse, error)).handleResponse(continuation: continuation)
+                    }
+                }
+            } else {
+                throw fatalError()
+            }
+        }
+    }
+    
+    /// Asynchronously fetch the current page
+    public var current: any PaginatedResult<T> {
+        get async throws {
+            return try await withCheckedThrowingContinuation { continuation in
+                continuation.resume(returning: self)
+            }
+        }
+    }
+}
+
+// TODO: Encode these instead
+extension QueryOptions {
+    func toDictionary() -> [String: String] {
```

Comment:
This is a vague name; a dictionary that represents what?

Reply:
it's an extension so it's only ever called on QueryOptions e.g. `queryOptions.toDictionary()` i.e. a dictionary representation of the query options

I can explicitly name it `toStringDictionary` but I don't think that's significantly better, wdyt?

Reply:
Exactly — it’s _a_ dictionary representation of the query options (emphasis mine). But there are many different ways in which you could represent a `QueryOptions` as a dictionary, for example depending on how you choose to map the `orderBy` property to a string, so the name of this function should help us understand what is special about _this_ representation. It seems like in this case this dictionary representation is going to be used as the params for a request to the Chat API, so how about `toChatAPIRequestParams` or some variation thereof?

Reply:
I personally think it makes sense in regards to `chat-swift` - we only have 1 API (ChatAPI) and 1 way of mapping properties _at the moment_. Granted, we _could_ introduce other ways but it seems unlikely. I try to follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and the [rule of three ](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) when trying to decide if certain design/semantic decisions are worth the time _right now_ or if they can be deferred indefinitely

The `ARTDataQuery` link from your other comment helped give me context from ably-cocoa so I'll rename this to match with what we do in [there](https://github.com/ably/ably-cocoa/blob/63e6f001d06cb7defb6be92f87a831f920eaf8c1/Source/ARTDataQuery.m#L25). It's essentially exactly the same function

Reply:
OK, can you change the extension to not be `public` please?

Reply:
fixed - marking as resolved to merge (PR already approved by yourself too)